pkgs/by-name/co/comfyui/package.nix

```diff
@@ -0,0 +1,148 @@
+{ lib
+, python311
+, linkFarm
+, writers
+, writeTextFile
+, fetchFromGitHub
+, stdenv
+, symlinkJoin
+, config
+, gpuBackend ? (
+  if config.cudaSupport
+  then "cuda"
+  else if config.rocmSupport
+  then "rocm"
+  else "none"
+)
+, modelsPath ? "/var/lib/comfyui/models"
+, inputPath ? "/var/lib/comfyui/input"
+, outputPath ? "/var/lib/comfyui/output"
+, tempPath ? "/var/lib/comfyui/temp"
+, userPath ? "/var/lib/comfyui/user"
+, customNodes ? []
+, models ? {
+    checkpoints = {};
+    clip = {};
+    clip_vision = {};
+    configs = {};
+    controlnet = {};
+    embeddings = {};
+    upscale_modules = {};
+    vae = {};
+    vae_approx = {};
+  }
+}:
+
+let
+
+  config-data = {
+    comfyui = {
+      base_path = modelsPath;
+      checkpoints = "${modelsPath}/checkpoints";
+      clip = "${modelsPath}/clip";
+      clip_vision = "${modelsPath}/clip_vision";
+      configs = "${modelsPath}/configs";
+      controlnet = "${modelsPath}/controlnet";
+      embeddings = "${modelsPath}/embeddings";
+      loras = "${modelsPath}/loras";
+      upscale_models= "${modelsPath}/upscale_models";
+      vae = "${modelsPath}/vae";
+      vae_approx = "${modelsPath}/vae_approx";
+    };
+  };
+
+  modelPathsFile = writeTextFile {
+    name = "extra_model_paths.yaml";
+    text = (lib.generators.toYAML {} config-data);
+  };
+
+  pythonEnv = (python311.withPackages (ps: with ps; [
+    (
+      if gpuBackend == "cuda"
+      then torchWithCuda
+      else if gpuBackend == "rocm"
+      then torchWithRocm
+      else torch
+    )
+    torchsde
+    torchvision
+    torchaudio
+    transformers
+    safetensors
+    accelerate
+    aiohttp
+    einops
+    kornia
+    pyyaml
+    pillow
+    scipy
+    psutil
+    tqdm
+  ] ++ (builtins.concatMap (node: node.dependencies) customNodes)));
+
+  executable = writers.writeDashBin "comfyui" ''
+    cd $out && \
+    ${pythonEnv}/bin/python comfyui \
+      --input-directory ${inputPath} \
+      --output-directory ${outputPath} \
+      --extra-model-paths-config ${modelPathsFile} \
+      --temp-directory ${tempPath} \
+      "$@"
+  '';
+
+  customNodesCollection = (
+    linkFarm "comfyui-custom-nodes" (builtins.map (pkg: { name = pkg.pname; path = pkg; }) customNodes)
+  );
+in stdenv.mkDerivation rec {
```

Comment:
I think we should use the `finalAttrs` pattern here.

The `finalAttrs` pattern will let you remove the `rec` keyword ([see its implementation in Nix](https://github.com/NixOS/nix/blob/2d1d81114d72ace89ce08cd3bc93f4eb27a2975d/src/libexpr/nixexpr.cc#L121)).

Why is this pattern preferred to `rec` ?

Let's take this simple code example:

```nix
mkDerivation rec {
   foo = 1;
   bar = foo + 1;
}
```

and then `.overrideAttrs(old: { foo = 2; })`, you'll get `{ foo = 2; bar = 2; }` while with `finalAttrs` pattern, it would work correctly because it's a real fixed point.

Let me share a couple of useful links regarding the `finalAttrs` pattern:

1. History: https://discourse.nixos.org/t/avoid-rec-expresions-in-nixpkgs/8293
2. Documentation: https://nixos.org/manual/nixpkgs/unstable/#mkderivation-recursive-attributes
3. Recent example of implementation: https://github.com/NixOS/nixpkgs/compare/17f96f7b978e61576cfe16136eb418f74fab9952..9e6ea843e473d34d4f379b8b0d8ef0425a06defe

Feel free to reach out if you need some assistance.